packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol

```diff
@@ -0,0 +1,217 @@
+    function claimWindows(
+        Claim[] memory claims,
+        address rewardToken,
+        address account
+    ) public nonReentrant() {
+        uint256 amountToClaim = 0;
+        for (uint256 i = 0; i < claims.length; i++) {
+            Claim memory claim = claims[i];
+            require(claim.account == account, "Invalid account in batch claim");
+            _markClaimed(claim);
+            amountToClaim = amountToClaim.add(claim.amount);
+        }
+        _disburse(IERC20(rewardToken), account, amountToClaim);
+    }
+
+    // Claim `amount` of reward tokens for `account`. If `amount` and `account` do not exactly match the values stored
+    // in the merkle proof for this `windowIndex` this method will revert.
+    function claimWindow(Claim memory claim) public nonReentrant() {
+        _markClaimed(claim);
+        _disburse(merkleWindows[claim.windowIndex].rewardToken, claim.account, claim.amount);
+    }
+
+    /****************************
+     *
+     * Internal functions
+     *
+     ****************************/
+
+    function _seedWindow(
+        uint256 windowIndex,
+        uint256 totalRewardsDistributed,
+        uint256 windowStart,
+        address rewardToken,
+        bytes32 merkleRoot
+    ) private {
+        Window storage window = merkleWindows[windowIndex];
+        window.start = windowStart;
+        window.merkleRoot = merkleRoot;
+        window.rewardToken = IERC20(rewardToken);
+
+        window.rewardToken.safeTransferFrom(msg.sender, address(this), totalRewardsDistributed);
+
+        emit SeededWindow(windowIndex, totalRewardsDistributed, windowStart, rewardToken, msg.sender);
+    }
+
+    function _markClaimed(Claim memory claim) private {
+        // Check claimed proof against merkle window at given index.
+        require(_verifyClaim(claim), "Incorrect merkle proof");
+        // Check the account has not yet claimed for this window.
+        require(!claimed[claim.windowIndex][claim.account], "Account has already claimed for this window");
+
+        // Proof is correct and claim has not occurred yet; check that claim window has begun.
+        require(getCurrentTime() >= merkleWindows[claim.windowIndex].start, "Claim window has not begin");
+
+        claimed[claim.windowIndex][claim.account] = true;
+        emit Claimed(
+            msg.sender,
+            claim.windowIndex,
+            claim.account,
+            claim.amount,
+            address(merkleWindows[claim.windowIndex].rewardToken)
+        );
+    }
+
+    function _disburse(
+        IERC20 token,
+        address account,
+        uint256 amount
+    ) private {
+        // TODO: Should we revert claims for 0 tokens?
+        if (amount > 0) {
+            token.safeTransfer(account, amount);
+        }
+    }
+
+    // Checks {account, amount} against Merkle root at given window index.
+    function _verifyClaim(Claim memory claim) private view returns (bool valid) {
```

Comment:
Was this previously public? I could see it making sense to keep it public 🤷‍♀️

Reply:
Yeah, but I don't really see a reason to keep it public vs private so by default I lean private. I think you could achieve the same "sanity" check by `.call()`-ing on `claimWindow`

Reply:
That's fair, although there is one minor difference -- if you've already claimed this would return true but claimWindow would revert. Not sure that's a huge deal because you can just || between the mapping and the `.call` to see if that person was included in a merkle root. Not strongly opinionated, and not sure we'd ever really need to do this.

Reply:
Yeah that confusion alone is worth removing for me.